File: scripts/peaks_annotation/peaks_annotation_NSC_combined.py
import pandas as pd
import matplotlib.pyplot as plt
from gtfparse import read_gtf
import numpy as np
from typing import Tuple, List
import seaborn as sns
import os
import argparse

def load_gencode_annotations(gtf_path: str) -> pd.DataFrame:
    """Load and process GENCODE GTF file."""
    print("Loading GENCODE annotations...")
    df_gtf = read_gtf(gtf_path)
    # Convert to pandas if it's a polars dataframe
    if hasattr(df_gtf, 'to_pandas'):
        df_gtf = df_gtf.to_pandas()
    # Filter for genes and basic annotations
    genes = df_gtf[df_gtf["feature"] == "gene"].copy()
    return genes

def categorize_region(position: Tuple[str, int, int], 
                     genes: pd.DataFrame,
                     promoter_size: int = 2000) -> str:
    """
    Categorize a genomic region based on its position relative to genes.
    """
    chrom, start, end = position
    
    # Filter genes for the relevant chromosome
    chr_genes = genes[genes["seqname"] == chrom]
    
    if chr_genes.empty:
        return "Distal Intergenic"
    
    for _, gene in chr_genes.iterrows():
        gene_start = gene["start"]
        gene_end = gene["end"]
        strand = gene["strand"]
        
        # Define promoter region based on strand
        if strand == "+":
            promoter_start = gene_start - promoter_size
            promoter_end = gene_start
        else:
            promoter_start = gene_end
            promoter_end = gene_end + promoter_size
            
        # Check if peak overlaps with promoter
        if max(start, promoter_start) <= min(end, promoter_end):
            return "Promoter"
        
        # Check if peak overlaps with gene body
        if max(start, gene_start) <= min(end, gene_end):
            # For simplicity, we're considering the first exon/intron as one category
            if (strand == "+" and abs(start - gene_start) <= 1000) or \
               (strand == "-" and abs(end - gene_end) <= 1000):
                return "1st Intron/Exon"
            return "Gene body"
            
    return "Distal Intergenic"

def load_narrowpeak(file_path: str) -> pd.DataFrame:
    """Load narrowPeak format file."""
    columns = ['chr', 'start', 'end', 'name', 'score', 
              'strand', 'signalValue', 'pValue', 'qValue', 'peak']
    return pd.read_csv(file_path, sep='\t', names=columns)

def analyze_peaks(peaks_file: str, genes: pd.DataFrame) -> dict:
    """Analyze peaks and categorize them into genomic regions."""
    print(f"Analyzing peaks from {peaks_file}...")
    peaks = load_narrowpeak(peaks_file)
    
    categories = {
        "Promoter": 0,
        "1st Intron/Exon": 0,
        "Gene body": 0,
        "5' UTR": 0,
        "3' UTR": 0,
        "Distal Intergenic": 0
    }
    
    for _, peak in peaks.iterrows():
        category = categorize_region(
            (peak["chr"], peak["start"], peak["end"]),
            genes
        )
        categories[category] += 1
    
    return categories

def plot_distributions(endo_data: dict, exo_data: dict, output_dir: str):
    """Create side-by-side pie charts for NPCs endogenous vs exogenous."""
    fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(15, 7))
    
    colors = {
        "Promoter": "#e5e5b5",
        "1st Intron/Exon": "#b5d7e5",
        "Gene body": "#3182bd",
        "5' UTR": "#e6550d",
        "3' UTR": "#fd8d3c",
        "Distal Intergenic": "#e7969c"
    }
    
    # Plot data
    datasets = [
        (endo_data, "NPCs Endogenous", ax1),
        (exo_data, "NPCs Exogenous", ax2)
    ]
    
    for data, title, ax in datasets:
        values = list(data.values())
        labels = list(data.keys())
        colors_list = [colors[label] for label in labels]
        
        ax.pie(values, labels=labels, colors=colors_list,
               autopct='%1.1f%%', startangle=90)
        ax.set_title(title)
    
    plt.savefig(os.path.join(output_dir, "peak_distributions_NPCs_combined.png"), 
                dpi=300, bbox_inches='tight')
    plt.close()

def main():
    parser = argparse.ArgumentParser(description='Analyze gene annotations in peaks.')
    parser.add_argument('--gtf-path', required=True, help='Path to GENCODE GTF file')
    parser.add_argument('--peaks-dir', required=True, help='Directory containing narrowPeak files')
    parser.add_argument('--output-dir', required=True, help='Output directory for plots')
    args = parser.parse_args()
    
    # Load gene annotations
    genes = load_gencode_annotations(args.gtf_path)
    
    # Create output directory
    os.makedirs(args.output_dir, exist_ok=True)
    
    # Define peak files
    peak_files = {
        'endo': os.path.join(args.peaks_dir, "NPCs_endo_combined.narrowPeak"),
        'exo': os.path.join(args.peaks_dir, "NPCs_exo_combined.narrowPeak")
    }
    
    # Process files
    results = {}
    for condition, file_path in peak_files.items():
        if os.path.exists(file_path):
            results[condition] = analyze_peaks(file_path, genes)
        else:
            print(f"Warning: File not found - {file_path}")
    
    # Create plot if both files were processed successfully
    if len(results) == 2:
        plot_distributions(results['endo'], results['exo'], args.output_dir)
        
        # Print statistics
        for condition, categories in results.items():
            print(f"\nDistribution for {condition}:")
            total = sum(categories.values())
            for category, count in categories.items():
                percentage = (count / total) * 100
                print(f"{category}: {count} peaks ({percentage:.1f}%)")

if __name__ == "__main__":
    main() 

File: scripts/peaks_annotation/peaks_annotation_NSC_combined.R
#!/usr/bin/env Rscript

# Load required packages
suppressPackageStartupMessages({
    library(optparse)
    library(ChIPseeker)
    library(GenomicFeatures)
    library(TxDb.Mmusculus.UCSC.mm10.knownGene)
    library(org.Mm.eg.db)
    library(rtracklayer)
    library(ggplot2)
})

# Parse command line arguments
option_list <- list(
    make_option("--peaks-dir", type="character", default=NULL, dest="peaks_dir",
                help="Directory containing narrowPeak files"),
    make_option("--output-dir", type="character", default=NULL, dest="output_dir",
                help="Output directory for plots and results"),
    make_option("--peak-type", type="character", default="narrow", dest="peak_type",
                help="Type of peaks to process: narrow or broad")
)
opts <- parse_args(OptionParser(option_list=option_list))

# Add diagnostic prints
print("Command line arguments received:")
print(opts)
print("Output directory path:")
print(opts$output_dir)
print("Peak type:")
print(opts$peak_type)

# Ensure the peak type is valid
if (!opts$peak_type %in% c("narrow", "broad")) {
    stop("Invalid peak type. Must be 'narrow' or 'broad'.")
}

# Ensure the output directory path is not NULL and is a character string
if (is.null(opts$output_dir) || !is.character(opts$output_dir)) {
    stop("Invalid or missing output directory path")
}

# Create output directory
dir.create(opts$output_dir, recursive = TRUE, showWarnings = FALSE)

# Main function
main <- function() {
    # Load TxDb
    message("Loading mm10 annotations...")
    txdb <- TxDb.Mmusculus.UCSC.mm10.knownGene
    
    # Process peak files
    peak_files <- list(
        endo = file.path(opts$peaks_dir, paste0("NPCs_endo_combined.", opts$peak_type, "Peak")),
        exo = file.path(opts$peaks_dir, paste0("NPCs_exo_combined.", opts$peak_type, "Peak"))
    )
    
    # Store annotations
    annotations <- list()
    
    # Process each peak file
    for (condition in names(peak_files)) {
        if (file.exists(peak_files[[condition]])) {
            message(sprintf("Processing %s peaks...", condition))
            
            # Read the peaks file manually
            peaks_df <- read.table(peak_files[[condition]], 
                                 col.names=c("chrom", "start", "end", "name", 
                                           "score", "strand", "signalValue",
                                           "pValue", "qValue", "peak"),
                                 colClasses=c("character", "numeric", "numeric", 
                                            "character", "numeric", "character",
                                            "numeric", "numeric", "numeric", "numeric"))
            
            # Convert "." in strand to "*"
            peaks_df$strand[peaks_df$strand == "."] <- "*"
            
            # Convert to GRanges
            peaks <- GRanges(seqnames=peaks_df$chrom,
                           ranges=IRanges(start=peaks_df$start, end=peaks_df$end),
                           name=peaks_df$name,
                           score=peaks_df$score,
                           strand=peaks_df$strand,
                           signalValue=peaks_df$signalValue,
                           pValue=peaks_df$pValue,
                           qValue=peaks_df$qValue,
                           peak=peaks_df$peak)
            
            # Annotate peaks
            annotations[[condition]] <- annotatePeak(peaks,
                                                   TxDb = txdb,
                                                   annoDb = "org.Mm.eg.db",
                                                   tssRegion = c(-3000, 3000),
                                                   level = "transcript")
            
            # Generate individual plots
            pdf(file.path(opts$output_dir, sprintf("%s_annotation_plots.pdf", condition)), 
                width = 10, height = 12)
            
            # Plot genomic annotation
            print(plotAnnoBar(annotations[[condition]]))
            
            # Plot distance to TSS
            print(plotDistToTSS(annotations[[condition]]))
            
            # Plot annotation pie chart
            print(plotAnnoPie(annotations[[condition]]))
            
            dev.off()
            
            # Save detailed annotation results
            anno_df <- as.data.frame(annotations[[condition]]@anno)
            write.table(anno_df,
                       file = file.path(opts$output_dir, 
                                      sprintf("%s_peak_annotations.tsv", condition)),
                       sep = "\t", quote = FALSE, row.names = FALSE)
        } else {
            warning(sprintf("File not found: %s", peak_files[[condition]]))
        }
    }
    
    # Create combined visualization if both conditions are present
    if (length(annotations) == 2) {
        pdf(file.path(opts$output_dir, "combined_annotation_plots.pdf"), 
            width = 12, height = 8)
        
        # Combined bar plot
        print(plotAnnoBar(annotations))
        
        # Combined distance to TSS
        print(plotDistToTSS(annotations))
        
        dev.off()
    }
}

# Run main function
main() 